src/tests/ftest/daos_test/suite.yaml

```diff
@@ -27,7 +27,7 @@ timeouts:
   test_daos_extend_simple: 3600
   test_daos_oid_allocator: 640
   test_daos_checksum: 500
-  test_daos_rebuild_ec: 6400
+  test_daos_rebuild_ec: 7200
```

Comment:
Just pointing out that a 2-hour timeout is very expensive for a PR test.

Reply:
this is devastating for PRs especially for ones that don't do with rebuild or EC.
my suggestion is to run a subset of these on PR (20 mins max) and the rest daily or weekly.

Reply:
I would rather to all these tests for each related PR.  Though the current PR test-set indeed include too much tests. So instead of moving the test cases outside of PR, I would suggest 

1. Split the PR test cases into several sub cases, for example control-panel, core-test, tests, utils, etc.

2. Each commit have to specify the test-tag or feature in its commit message, it is up to the reviewer or module owner decide if the test-tag/feature is enough.  Then the CI test case will only run those tests specified by test-tag.  Currently, each commit will run all PR tests by default.  I saw a lot control panel or test modification PR just all of PR tests, which might be the main reason for the slowness of turn-around.

3. Setup a staging phase and assign a special gatekeeper, so after each PR pass review and its CI test cases, then the PR will be landed to staging branch,  then on each Friday, the gatekeeper will run weekly test on the staging branch, if some tests failed, the special gatekeeper can revert the related PR to tell the author, otherwise land all left PRs to the real master branch.  This should largely avoid the regression.

what do you think about this process? @mchaarawi @daltonbohning   

we discussed about this process before with @gnailzenh  @NiuYawei @liuxuezhao

Reply:
> Currently, each commit will run all PR tests by default. I saw a lot control panel or test modification PR just all of PR tests, which might be the main reason for the slowness of turn-around.

If `test_daos_rebuild_ec` shouldn't run on all PRs then it shouldn't be tagged with `pr`.
`pr` tests are meant to be run for every PR. Then `Features: foo` will include `foo` tests, in addition to running `pr`.

Reply:
If all PRs ran and passed with `Features` then we would have few regressions in master. So there would be no need for a special staging branch. And what if there are 10 commits on the special staging branch and a test fails? How do you know which commit introduced the regression? With long CI turnaround, we could spend another week trying to fix the regression on the staging branch. Which holds up everyone's PRs just because one PR didn't run with `Features`.

But if there is a need for a staging branch, that's exactly what feature branches are for.

Reply:
yea let's review again with Liang and Xuezhao and Niu when they are back.
as we reviewed in gatekeeping last time, EC and rebuild dominate the PR test time. 
As Dalton said if we just have some basic tests for those on all PRs that run in a matter of few minutes, and for actual PRs that need to run the rebuild and EC tests you add the Features pragma, then everyone wins :-)
CI time would be much better.

Reply:
>And what if there are 10 commits on the special staging branch and a test fails? How do you know which commit introduced the regression? With long CI turnaround, we could spend another week trying to fix the regression on the staging branch. Which holds up everyone's PRs just because one PR didn't run with Features.

With limited commit per week, I do not think it is difficult to figure out which PR causing the failure.  I assume without competition, the turnaround time would not be that long.  Well, we only need simply revert the PR which causing the failure, to make sure the master is clean. why would need another week?

>as we reviewed in gatekeeping last time, EC and rebuild dominate the PR test time.
As Dalton said if we just have some basic tests for those on all PRs that run in a matter of few minutes, and for 

hmm rebuild & EC cost about 4 hours? Yes, we can split the rebuild & EC test cases,  but without staging area, I am  a bit concern, since all daos core changes might impact rebuild.  I would rather to make sure all PRs go through weekly/dayly tests before go into master.

Reply:
>But if there is a need for a staging branch, that's exactly what feature branches are for.

My points are

1. we need shrink the CI time generally for each PR, so it can move out of developer hands and  "commit" as soon as possible.  So we split the PR tests, and only run necessary and good enough tests.

2. To make sure no regression in master, we must run all weekly/daily tests for all PRs and also sequentially, so we need staging phase.

Feature branch can not achieve these, IMHO

Reply:
We run weekly testing once per week already on master. Landing to a staging branch and running the same weekly testing on the staging branch means the staging branch will be just as unstable as master is now. And if every PR has to go to the staging branch first, then every PR is still being affected by the exact same instability.

Reply:
> With limited commit per week, I do not think it is difficult to figure out which PR causing the failure. I assume without competition, the turnaround time would not be that long. Well, we only need simply revert the PR which causing the failure, to make sure the master is clean. why would need another week?

For reference, this is a recent weekly-testing run. Just weekly, not daily
https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/weekly-testing/366/pipeline/81
HW Large alone took almost 19 hours, not including queue time.
HW Large MD on SSD took another 19 hours.
HW Medium took 21 hours.
HW Medium MD on SSD took almost 22 hours.

So if we had a staging branch and needed to revert a commit, we would have to wait an entire day to get results again. And with only (I think) 2 HW Large clusters, and (I think) 3 HW Medium clusters, the staging branch would block progress on all other PRs until we merge the staging branch into master.

The problem is that we simply have too many tests and too many variants for the CI resources.

Reply:
And a reference daily testing
https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/daily-testing/68/pipeline/83
HW Large is < 1 hour.
HW Medium took almost 11 hours.

There is a little overlap between daily and weekly, but not much.
So adding them together, HW Medium would have a turnaround time of ~30 hours to run weekly + daily

Reply:
>So if we had a staging branch and needed to revert a commit, we would have to wait an entire day to get results again. And with only (I think) 2 HW Large clusters, and (I think) 3 HW Medium clusters, the staging branch would block progress on all other PRs until we merge the staging branch into master.

Most of the time, it might not need revert PR or only at most revert 1 PR, otherwise it means our "small" acceptance test suite is not good enough, then we need improve the test suite for each feature.   And also if we have staging phase, then weekly test on master is not needed anymore, maybe only monthly or as required.

The current process just easily causes the regression on master, IMHO.

Reply:
> With limited commit per week, I do not think it is difficult to figure out which PR causing the failure.

To put this in context it's three months since we've had a daily build passing all tests and we've never had a weekly build passing that I'm aware of.  I understand that in theory it should be easy to identify the cause of regressions and revert/fix as required but in practice that's something that we as a team haven't been very good at.

Reply:
> And also if we have staging phase, then weekly test on master is not needed anymore, maybe only monthly or as required.
>
> The current process just easily causes the regression on master, IMHO.

If I understand correctly, you propose creating a staging branch  - let's call it `staging`. Then all PRs that previously would have been landed to master would instead land to `staging`. Then, once per week we run the full daily and weekly test branch on `staging`. And if everything passes / no new regressions we merge `staging` into master. And to make CI time for running the suite on `staging`, we stop running weekly every week on `master`.

We currently run daily tests on master _**once per day**_, and we run weekly testing on master _**once per week**_. And master is unstable.

If we run daily testing on `staging` _**once per week**_ and weekly testing on `staging` _**once per week**_, that is a reduction in test coverage compared to what we run on master today.

Yes, master will be more stable. But that doesn't matter because PRs have to land to `staging` first. So effectively all this does is rename `master` -> `staging` and make the merge process more complicated.

Reply:
The reason we have staging branch

1. every PR will run sequentially with the same base, which will largely avoid the regression due to the conflict. Since otherwise, each PR will run CI based on different top.
2. I think would be easier to find the culprit and revert it, if there are any regression.
3. We have a clean master in any case.  It would save a lot time for each developer.

Reply:
that staging branch sounds like a nightmare situation tbh in terms of both management on when/how we move changes to master, and CI time having to test both master and that branch. no it's not acceptable in that case to test one but not the other.
We are complicating the situation too much tbh. we should just reduce the amount of tests for rebuild on every PR, and the PRs that touch rebuild need to add a Features: tag to just run more test. this solution is very simple and will improve the quality of life for everyone! :-)

Reply:
TBH, I thought our current process just let too many regression to get in master too easy, and even removing some of rebuild tests might only save you like 2 hours on each case.  So we have to split the test in more subtle category for each PR to save time.  To avoid regression,  we have to let every PR go through all test cases and sequentially(instead of basing on every different top), Note: I do not mean run each PR one by one, I mean run all PRs in staging branch all together by gatekeeper.

>HW Large alone took almost 19 hours, not including queue time.
HW Large MD on SSD took another 19 hours.
HW Medium took 21 hours.
HW Medium MD on SSD took almost 22 hours.

Reply:
2 hours out of 4 on PR tests is a HUGE improvement.
im not talking about weekly or daily test, just PR. 
TBH if current process lets regressions in easily, then the tests for that regression areas is not done properly and has to be revisited and always required to use a Features: tag for that PR.

Reply:
>2 hours out of 4 on PR tests is a HUGE improvement.

I thought 2 hours can only be saved in one subtest (rebuild ec in hardware medium, though I did not check how long other test cases will cost there, I am a bit skeptical about how long we can save by only removing rebuild test, let along we can remove all of it. @daltonbohning  any detail information here? ).

> im not talking about weekly or daily test, just PR.

I think we are at the same page on this one. the current PR tests include so much test cases. we need shrink it from some PR.

> TBH if current process lets regressions in easily, then the tests for that regression areas is not done properly and has to be revisited and always required to use a Features: tag for that PR.

TBH, I think all core changes need to run rebuild tests, even cart changes. PR-13571 is an good example here. Having these PRs run all weekly tests in group would largely avoid regression and saving time as well.

Reply:
> I thought 2 hours can only be saved in one subtest (rebuild ec in hardware medium, though I did not check how long other test cases will cost there, I am a bit skeptical about how long we can save by only removing rebuild test, let along we can remove all of it. @daltonbohning any detail information here? ).

This is a PR that just ran `pr` tests
https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-13752/4/pipeline

HW Large took ~30 minutes.
HW Medium took ~4 hours.
HW Medium Verbs took ~4 hours.

`test_daos_rebuild_ec` took ~1.5 hours on HW Medium Verbs:
https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-13752/4/testReport/FTEST_daos_test/DaosCoreTest-DAOS_Rebuild_EC/

So this ONE test accounted for 1.5 / 8 hours (19%) of HW Medium time.

Reply:
> TBH, I think all core changes need to run rebuild tests, even cart changes. PR-13571 is an good example here. Having these PRs run all weekly tests in group would largely avoid regression and saving time as well.

If we had the resources, all PRs would run all tests. But we literally don't have enough CI resources for all PRs to run all rebuild tests.

Resource constraints are the main reason why tests are bucketed into pr, daily, and weekly. The idea is
- pr - quick smoke/sanity basic tests that all PRs should run. This can include one or more tests from each area of code in  DAOS
- daily - Tests that are less critical than pr, take too long to run in pr, or generally aren't affected by "just any change"
- weekly - Same reasoning as daily, except less critical or longer running

And regardless of whether a test is pr, daily, or weekly, using `Features: foo` will run all tests tagged with `foo`. So the suggestion here is that `test_daos_rebuild_ec` (and any large test or feature area) should be split:
1. A relatively quick variant to run in pr
2. Some variant(s) in daily/weekly depending on how critical or long-running


Then, for example, if a PR modifies rebuild code it should use `Features: rebuild` to run all rebuild tests. And similar for other code areas. Of course, this doesn't ensure regressions don't slip in, but we _**cannot**_ run all tests on all PRs. So it's a matter of statistics: changes to rebuild code are more likely to affect rebuild than changes to DFS for example.

And even if ALL PRs ran ALL tests and passed, there is still the possibility of intermittent failures or breakage due to dependency updates, network changes/interruptions, etc.

Reply:
> Then, for example, if a PR modifies rebuild code it should use `Features: rebuild` to run all rebuild tests. And similar for other code areas. Of course, this doesn't ensure regressions don't slip in, but we _**cannot**_ run all tests on all PRs. So it's a matter of statistics: changes to rebuild code are more likely to affect rebuild than changes to DFS for example.

Oh, I agree with you on this one.  we do not have enough resource to let all PR run all tests individually.  that is why I suggest to run all PR in group by gatekeeper before landing to master.  I understood this will definitely not 100% avoid regression, but largely avoiding it.  

On the other hand, if we keeps shrink the test case for each PR, and then let them get into master with less coverage test, which likely make things worse, IMHO.

> And even if ALL PRs ran ALL tests and passed, there is still the possibility of intermittent failures or breakage due to dependency updates, network changes/interruptions, etc.

Totally agree, but as I see, most regression are actually due to lack of coverage and landing conflict, like DAOS-14985.

Reply:
> that is why I suggest to run all PR in group by gatekeeper before landing to master. I understood this will definitely not 100% avoid regression, but largely avoiding it.

I fail to see the logic here as I've already stated: https://github.com/daos-stack/daos/pull/13765#discussion_r1486494687

Reply:
> Oh, I agree with you on this one. we do not have enough resource to let all PR run all tests individually. that is why I suggest to run all PR in group by gatekeeper before landing to master. I understood this will definitely not 100% avoid regression, but largely avoiding it.

What you're describing here is the daily-testing job.  It's for important tests that provide good coverage but are otherwise too expensive to run for every build.  It's daily so we can swiftly identify regressions from a small set of patches and revert if appropriate.  Where PRs touch an area of code and require additional testing before landing then the Features tag should be used.

Reply:
>I fail to see the logic here as I've already stated: https://github.com/daos-stack/daos/pull/13765#discussion_r1486494687

Sorry, which logic you mean?  this one
>Yes, master will be more stable. But that doesn't matter because PRs have to land to staging first. So effectively all this does is rename master -> staging and make the merge process more complicated.

I think 
1. everyone will have a stable master to base on, which will save a lot time, no need re-trigger, no need spending a lot time to find out if this is known issue.
2. And also if gatekeeper control the batch commit, then all PR will be tested based on the same master, then no conflict and also all PR will run all tests before landing to master.

Reply:
> > Oh, I agree with you on this one. we do not have enough resource to let all PR run all tests individually. that is why I suggest to run all PR in group by gatekeeper before landing to master. I understood this will definitely not 100% avoid regression, but largely avoiding it.
> 
> What you're describing here is the daily-testing job. It's for important tests that provide good coverage but are otherwise too expensive to run for every build. It's daily so we can swiftly identify regressions from a small set of patches and revert if appropriate. Where PRs touch an area of code and require additional testing before landing then the Features tag should be used.

I think I am saying
1. the current PRs are landed to master without enough tests. IMHO, it is almost impossible to avoid the regression without letting all PRs go through all of tests.  
2. the current PR tests include so much tests, which might be irrelevant for some patches.  For example control panel change might only need run local + vm tests,  dfuse changes might not need hardware test at all. So I mean all patches can run less coverage tests before getting into staging, which saves both testing resource and developer time.   Then all PR needs to go through weekly tests before getting into master.  Though we would need a dedicate gatekeeper here.
3. Sue the current daily, weekly tests can identify the regression, but it can not maintain a clean master, then the regression becomes everyone's problem.